// @aries-framework/core/build/modules/cache/singleContextLruCache/SingleContextStorageLruCache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SingleContextStorageLruCache = void 0;
const lru_map_1 = require("lru_map");
const error_1 = require("../../../error");
const SingleContextLruCacheRecord_1 = require("./SingleContextLruCacheRecord");
const SingleContextLruCacheRepository_1 = require("./SingleContextLruCacheRepository");
const CONTEXT_STORAGE_LRU_CACHE_ID = 'CONTEXT_STORAGE_LRU_CACHE_ID';
/**
 * Cache that leverages the storage associated with the agent context to store cache records.
 * It will keep an in-memory cache of the records to avoid hitting the storage on every read request.
 * Therefor this cache is meant to be used with a single instance of the agent.
 *
 * Due to keeping an in-memory copy of the cache, it is also not meant to be used with multiple
 * agent context instances (meaning multi-tenancy), as they will overwrite the in-memory cache.
 *
 * However, this means the cache is not meant for usage with multiple instances.
 */
class SingleContextStorageLruCache {
    constructor({ limit }) {
        this.limit = limit;
    }
    async get(agentContext, key) {
        this.assertContextCorrelationId(agentContext);
        const cache = await this.getCache(agentContext);
        this.removeExpiredItems(cache);
        const item = cache.get(key);
        // Does not exist
        if (!item)
            return null;
        // Expired
        if (item.expiresAt && Date.now() > item.expiresAt) {
            cache.delete(key);
            await this.persistCache(agentContext);
            return null;
        }
        return item.value;
    }
    async set(agentContext, key, value, expiresInSeconds) {
        this.assertContextCorrelationId(agentContext);
        let expiresDate = undefined;
        if (expiresInSeconds) {
            expiresDate = new Date();
            expiresDate.setSeconds(expiresDate.getSeconds() + expiresInSeconds);
        }
        const cache = await this.getCache(agentContext);
        this.removeExpiredItems(cache);
        cache.set(key, {
            expiresAt: expiresDate === null || expiresDate === void 0 ? void 0 : expiresDate.getTime(),
            value,
        });
        await this.persistCache(agentContext);
    }
    async remove(agentContext, key) {
        this.assertContextCorrelationId(agentContext);
        const cache = await this.getCache(agentContext);
        this.removeExpiredItems(cache);
        cache.delete(key);
        await this.persistCache(agentContext);
    }
    async getCache(agentContext) {
        if (!this._cache) {
            const cacheRecord = await this.fetchCacheRecord(agentContext);
            this._cache = this.lruFromRecord(cacheRecord);
        }
        return this._cache;
    }
    lruFromRecord(cacheRecord) {
        return new lru_map_1.LRUMap(this.limit, cacheRecord.entries.entries());
    }
    async fetchCacheRecord(agentContext) {
        const cacheRepository = agentContext.dependencyManager.resolve(SingleContextLruCacheRepository_1.SingleContextLruCacheRepository);
        let cacheRecord = await cacheRepository.findById(agentContext, CONTEXT_STORAGE_LRU_CACHE_ID);
        if (!cacheRecord) {
            cacheRecord = new SingleContextLruCacheRecord_1.SingleContextLruCacheRecord({
                id: CONTEXT_STORAGE_LRU_CACHE_ID,
                entries: new Map(),
            });
            await cacheRepository.save(agentContext, cacheRecord);
        }
        return cacheRecord;
    }
    removeExpiredItems(cache) {
        cache.forEach((value, key) => {
            if (value.expiresAt && Date.now() > value.expiresAt) {
                cache.delete(key);
            }
        });
    }
    async persistCache(agentContext) {
        const cacheRepository = agentContext.dependencyManager.resolve(SingleContextLruCacheRepository_1.SingleContextLruCacheRepository);
        const cache = await this.getCache(agentContext);
        await cacheRepository.update(agentContext, new SingleContextLruCacheRecord_1.SingleContextLruCacheRecord({
            entries: new Map(cache.toJSON().map(({ key, value }) => [key, value])),
            id: CONTEXT_STORAGE_LRU_CACHE_ID,
        }));
    }
    /**
     * Asserts this class is not used with multiple agent context instances.
     */
    assertContextCorrelationId(agentContext) {
        if (!this._contextCorrelationId) {
            this._contextCorrelationId = agentContext.contextCorrelationId;
        }
        if (this._contextCorrelationId !== agentContext.contextCorrelationId) {
            throw new error_1.AriesFrameworkError('SingleContextStorageLruCache can not be used with multiple agent context instances. Register a custom cache implementation in the CacheModule.');
        }
    }
}
exports.SingleContextStorageLruCache = SingleContextStorageLruCache;
