// @credo-ts/core/build/modules/dids/methods/peer/PeerDidResolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PeerDidResolver = void 0;
const error_1 = require("../../../../error");
const repository_1 = require("../../repository");
const didPeer_1 = require("./didPeer");
const peerDidNumAlgo0_1 = require("./peerDidNumAlgo0");
const peerDidNumAlgo2_1 = require("./peerDidNumAlgo2");
const peerDidNumAlgo4_1 = require("./peerDidNumAlgo4");
class PeerDidResolver {
    constructor() {
        this.supportedMethods = ['peer'];
        /**
         * No remote resolving done, did document is fetched from storage. To not pollute the cache we don't allow caching
         */
        this.allowsCaching = false;
    }
    async resolve(agentContext, did) {
        const didRepository = agentContext.dependencyManager.resolve(repository_1.DidRepository);
        const didDocumentMetadata = {};
        try {
            let didDocument;
            if (!(0, didPeer_1.isValidPeerDid)(did)) {
                throw new error_1.CredoError(`did ${did} is not a valid peer did`);
            }
            const numAlgo = (0, didPeer_1.getNumAlgoFromPeerDid)(did);
            // For method 0, generate from did
            if (numAlgo === didPeer_1.PeerDidNumAlgo.InceptionKeyWithoutDoc) {
                didDocument = (0, peerDidNumAlgo0_1.didToNumAlgo0DidDocument)(did);
            }
            // For Method 1, retrieve from storage
            else if (numAlgo === didPeer_1.PeerDidNumAlgo.GenesisDoc) {
                // We can have multiple did document records stored for a single did (one created and one received). In this case it
                // doesn't matter which one we use, and they should be identical. So we just take the first one.
                const [didDocumentRecord] = await didRepository.findAllByDid(agentContext, did);
                if (!didDocumentRecord) {
                    throw new error_1.CredoError(`No did record found for peer did ${did}.`);
                }
                if (!didDocumentRecord.didDocument) {
                    throw new error_1.CredoError(`Found did record for method 1 peer did (${did}), but no did document.`);
                }
                didDocument = didDocumentRecord.didDocument;
            }
            // For Method 2, generate from did
            else if (numAlgo === didPeer_1.PeerDidNumAlgo.MultipleInceptionKeyWithoutDoc) {
                didDocument = (0, peerDidNumAlgo2_1.didToNumAlgo2DidDocument)(did);
            }
            // For Method 4, if short form is received, attempt to get the didDocument from stored record
            else {
                if ((0, peerDidNumAlgo4_1.isShortFormDidPeer4)(did)) {
                    const [didRecord] = await didRepository.findAllByDid(agentContext, did);
                    if (!didRecord) {
                        throw new error_1.CredoError(`No did record found for peer did ${did}.`);
                    }
                    didDocument = (0, peerDidNumAlgo4_1.didToNumAlgo4DidDocument)(didRecord.did);
                }
                else {
                    didDocument = (0, peerDidNumAlgo4_1.didToNumAlgo4DidDocument)(did);
                }
            }
            return {
                didDocument,
                didDocumentMetadata,
                didResolutionMetadata: { contentType: 'application/did+ld+json' },
            };
        }
        catch (error) {
            return {
                didDocument: null,
                didDocumentMetadata,
                didResolutionMetadata: {
                    error: 'notFound',
                    message: `resolver_error: Unable to resolve did '${did}': ${error}`,
                },
            };
        }
    }
}
exports.PeerDidResolver = PeerDidResolver;
