// @credo-ts/core/build/modules/dids/methods/jwk/JwkDidResolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JwkDidResolver = void 0;
const DidJwk_1 = require("./DidJwk");
class JwkDidResolver {
    constructor() {
        this.supportedMethods = ['jwk'];
        /**
         * No remote resolving done, did document is dynamically constructed. To not pollute the cache we don't allow caching
         */
        this.allowsCaching = false;
    }
    async resolve(agentContext, did) {
        const didDocumentMetadata = {};
        try {
            const didDocument = DidJwk_1.DidJwk.fromDid(did).didDocument;
            return {
                didDocument,
                didDocumentMetadata,
                didResolutionMetadata: { contentType: 'application/did+ld+json' },
            };
        }
        catch (error) {
            return {
                didDocument: null,
                didDocumentMetadata,
                didResolutionMetadata: {
                    error: 'notFound',
                    message: `resolver_error: Unable to resolve did '${did}': ${error}`,
                },
            };
        }
    }
}
exports.JwkDidResolver = JwkDidResolver;
