// @credo-ts/core/build/modules/vc/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SECURITY_CONTEXT_SECP256k1_URL = exports.SECURITY_JWS_CONTEXT_URL = exports.EXPANDED_TYPE_CREDENTIALS_CONTEXT_V1_VC_TYPE = exports.VERIFIABLE_PRESENTATION_TYPE = exports.VERIFIABLE_CREDENTIAL_TYPE = exports.SECURITY_SIGNATURE_URL = exports.SECURITY_PROOF_URL = exports.CREDENTIALS_ISSUER_URL = exports.SECURITY_CONTEXT_BBS_URL = exports.CREDENTIALS_CONTEXT_V1_URL = exports.DID_V1_CONTEXT_URL = exports.SECURITY_X25519_CONTEXT_URL = exports.SECURITY_CONTEXT_URL = exports.SECURITY_CONTEXT_V3_URL = exports.SECURITY_CONTEXT_V2_URL = exports.SECURITY_CONTEXT_V1_URL = void 0;
exports.SECURITY_CONTEXT_V1_URL = 'https://w3id.org/security/v1';
exports.SECURITY_CONTEXT_V2_URL = 'https://w3id.org/security/v2';
exports.SECURITY_CONTEXT_V3_URL = 'https://w3id.org/security/v3-unstable';
exports.SECURITY_CONTEXT_URL = exports.SECURITY_CONTEXT_V2_URL;
exports.SECURITY_X25519_CONTEXT_URL = 'https://w3id.org/security/suites/x25519-2019/v1';
exports.DID_V1_CONTEXT_URL = 'https://www.w3.org/ns/did/v1';
exports.CREDENTIALS_CONTEXT_V1_URL = 'https://www.w3.org/2018/credentials/v1';
exports.SECURITY_CONTEXT_BBS_URL = 'https://w3id.org/security/bbs/v1';
exports.CREDENTIALS_ISSUER_URL = 'https://www.w3.org/2018/credentials#issuer';
exports.SECURITY_PROOF_URL = 'https://w3id.org/security#proof';
exports.SECURITY_SIGNATURE_URL = 'https://w3id.org/security#signature';
exports.VERIFIABLE_CREDENTIAL_TYPE = 'VerifiableCredential';
exports.VERIFIABLE_PRESENTATION_TYPE = 'VerifiablePresentation';
exports.EXPANDED_TYPE_CREDENTIALS_CONTEXT_V1_VC_TYPE = 'https://www.w3.org/2018/credentials#VerifiableCredential';
exports.SECURITY_JWS_CONTEXT_URL = 'https://w3id.org/security/suites/jws-2020/v1';
exports.SECURITY_CONTEXT_SECP256k1_URL = 'https://w3id.org/security/suites/secp256k1-2019/v1';
