// @credo-ts/core/build/modules/vc/W3cCredentialService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.W3cCredentialService = void 0;
const error_1 = require("../../error");
const plugins_1 = require("../../plugins");
const constants_1 = require("./constants");
const data_integrity_1 = require("./data-integrity");
const W3cJsonLdCredentialService_1 = require("./data-integrity/W3cJsonLdCredentialService");
const W3cJsonLdVerifiablePresentation_1 = require("./data-integrity/models/W3cJsonLdVerifiablePresentation");
const jwt_vc_1 = require("./jwt-vc");
const W3cJwtCredentialService_1 = require("./jwt-vc/W3cJwtCredentialService");
const models_1 = require("./models");
const W3cPresentation_1 = require("./models/presentation/W3cPresentation");
const repository_1 = require("./repository");
let W3cCredentialService = class W3cCredentialService {
    constructor(w3cCredentialRepository, w3cJsonLdCredentialService, w3cJwtCredentialService) {
        this.w3cCredentialRepository = w3cCredentialRepository;
        this.w3cJsonLdCredentialService = w3cJsonLdCredentialService;
        this.w3cJwtCredentialService = w3cJwtCredentialService;
    }
    /**
     * Signs a credential
     *
     * @param credential the credential to be signed
     * @returns the signed credential
     */
    async signCredential(agentContext, options) {
        if (options.format === models_1.ClaimFormat.JwtVc) {
            const signed = await this.w3cJwtCredentialService.signCredential(agentContext, options);
            return signed;
        }
        else if (options.format === models_1.ClaimFormat.LdpVc) {
            const signed = await this.w3cJsonLdCredentialService.signCredential(agentContext, options);
            return signed;
        }
        else {
            throw new error_1.CredoError(`Unsupported format in options. Format must be either 'jwt_vc' or 'ldp_vc'`);
        }
    }
    /**
     * Verifies the signature(s) of a credential
     */
    async verifyCredential(agentContext, options) {
        if (options.credential instanceof data_integrity_1.W3cJsonLdVerifiableCredential) {
            return this.w3cJsonLdCredentialService.verifyCredential(agentContext, options);
        }
        else if (options.credential instanceof jwt_vc_1.W3cJwtVerifiableCredential || typeof options.credential === 'string') {
            return this.w3cJwtCredentialService.verifyCredential(agentContext, options);
        }
        else {
            throw new error_1.CredoError(`Unsupported credential type in options. Credential must be either a W3cJsonLdVerifiableCredential or a W3cJwtVerifiableCredential`);
        }
    }
    /**
     * Utility method that creates a {@link W3cPresentation} from one or more {@link W3cJsonLdVerifiableCredential}s.
     *
     * **NOTE: the presentation that is returned is unsigned.**
     *
     * @returns An instance of {@link W3cPresentation}
     */
    async createPresentation(options) {
        const presentation = new W3cPresentation_1.W3cPresentation({
            context: [constants_1.CREDENTIALS_CONTEXT_V1_URL],
            type: ['VerifiablePresentation'],
            verifiableCredential: options.credentials,
            holder: options.holder,
            id: options.id,
        });
        return presentation;
    }
    /**
     * Signs a presentation including the credentials it includes
     *
     * @param presentation the presentation to be signed
     * @returns the signed presentation
     */
    async signPresentation(agentContext, options) {
        if (options.format === models_1.ClaimFormat.JwtVp) {
            const signed = await this.w3cJwtCredentialService.signPresentation(agentContext, options);
            return signed;
        }
        else if (options.format === models_1.ClaimFormat.LdpVp) {
            const signed = await this.w3cJsonLdCredentialService.signPresentation(agentContext, options);
            return signed;
        }
        else {
            throw new error_1.CredoError(`Unsupported format in options. Format must be either 'jwt_vp' or 'ldp_vp'`);
        }
    }
    /**
     * Verifies a presentation including the credentials it includes
     *
     * @param presentation the presentation to be verified
     * @returns the verification result
     */
    async verifyPresentation(agentContext, options) {
        if (options.presentation instanceof W3cJsonLdVerifiablePresentation_1.W3cJsonLdVerifiablePresentation) {
            return this.w3cJsonLdCredentialService.verifyPresentation(agentContext, options);
        }
        else if (options.presentation instanceof jwt_vc_1.W3cJwtVerifiablePresentation ||
            typeof options.presentation === 'string') {
            return this.w3cJwtCredentialService.verifyPresentation(agentContext, options);
        }
        else {
            throw new error_1.CredoError('Unsupported credential type in options. Presentation must be either a W3cJsonLdVerifiablePresentation or a W3cJwtVerifiablePresentation');
        }
    }
    /**
     * Writes a credential to storage
     *
     * @param record the credential to be stored
     * @returns the credential record that was written to storage
     */
    async storeCredential(agentContext, options) {
        let expandedTypes = [];
        // JsonLd credentials need expanded types to be stored.
        if (options.credential instanceof data_integrity_1.W3cJsonLdVerifiableCredential) {
            expandedTypes = await this.w3cJsonLdCredentialService.getExpandedTypesForCredential(agentContext, options.credential);
        }
        // Create an instance of the w3cCredentialRecord
        const w3cCredentialRecord = new repository_1.W3cCredentialRecord({
            tags: { expandedTypes },
            credential: options.credential,
        });
        // Store the w3c credential record
        await this.w3cCredentialRepository.save(agentContext, w3cCredentialRecord);
        return w3cCredentialRecord;
    }
    async removeCredentialRecord(agentContext, id) {
        await this.w3cCredentialRepository.deleteById(agentContext, id);
    }
    async getAllCredentialRecords(agentContext) {
        return await this.w3cCredentialRepository.getAll(agentContext);
    }
    async getCredentialRecordById(agentContext, id) {
        return await this.w3cCredentialRepository.getById(agentContext, id);
    }
    async findCredentialsByQuery(agentContext, query) {
        const result = await this.w3cCredentialRepository.findByQuery(agentContext, query);
        return result.map((record) => record.credential);
    }
    async findCredentialRecordByQuery(agentContext, query) {
        const result = await this.w3cCredentialRepository.findSingleByQuery(agentContext, query);
        return result === null || result === void 0 ? void 0 : result.credential;
    }
};
W3cCredentialService = __decorate([
    (0, plugins_1.injectable)(),
    __metadata("design:paramtypes", [repository_1.W3cCredentialRepository,
        W3cJsonLdCredentialService_1.W3cJsonLdCredentialService,
        W3cJwtCredentialService_1.W3cJwtCredentialService])
], W3cCredentialService);
exports.W3cCredentialService = W3cCredentialService;
