// @credo-ts/core/build/modules/proofs/protocol/BaseProofProtocol.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseProofProtocol = void 0;
const EventEmitter_1 = require("../../../agent/EventEmitter");
const didcomm_1 = require("../../../storage/didcomm");
const ProofEvents_1 = require("../ProofEvents");
const ProofState_1 = require("../models/ProofState");
const repository_1 = require("../repository");
class BaseProofProtocol {
    async processProblemReport(messageContext) {
        const { message: proofProblemReportMessage, agentContext, connection } = messageContext;
        agentContext.config.logger.debug(`Processing problem report with message id ${proofProblemReportMessage.id}`);
        const proofRecord = await this.getByProperties(agentContext, {
            threadId: proofProblemReportMessage.threadId,
            connectionId: connection === null || connection === void 0 ? void 0 : connection.id,
        });
        // Update record
        proofRecord.errorMessage = `${proofProblemReportMessage.description.code}: ${proofProblemReportMessage.description.en}`;
        await this.updateState(agentContext, proofRecord, ProofState_1.ProofState.Abandoned);
        return proofRecord;
    }
    /**
     * Update the record to a new state and emit an state changed event. Also updates the record
     * in storage.
     *
     * @param proofRecord The proof record to update the state for
     * @param newState The state to update to
     *
     */
    async updateState(agentContext, proofRecord, newState) {
        const proofRepository = agentContext.dependencyManager.resolve(repository_1.ProofRepository);
        agentContext.config.logger.debug(`Updating proof record ${proofRecord.id} to state ${newState} (previous=${proofRecord.state})`);
        const previousState = proofRecord.state;
        proofRecord.state = newState;
        await proofRepository.update(agentContext, proofRecord);
        this.emitStateChangedEvent(agentContext, proofRecord, previousState);
    }
    emitStateChangedEvent(agentContext, proofRecord, previousState) {
        const eventEmitter = agentContext.dependencyManager.resolve(EventEmitter_1.EventEmitter);
        eventEmitter.emit(agentContext, {
            type: ProofEvents_1.ProofEventTypes.ProofStateChanged,
            payload: {
                proofRecord: proofRecord.clone(),
                previousState: previousState,
            },
        });
    }
    /**
     * Retrieve a proof record by id
     *
     * @param proofRecordId The proof record id
     * @throws {RecordNotFoundError} If no record is found
     * @return The proof record
     *
     */
    getById(agentContext, proofRecordId) {
        const proofRepository = agentContext.dependencyManager.resolve(repository_1.ProofRepository);
        return proofRepository.getById(agentContext, proofRecordId);
    }
    /**
     * Retrieve all proof records
     *
     * @returns List containing all proof records
     */
    getAll(agentContext) {
        const proofRepository = agentContext.dependencyManager.resolve(repository_1.ProofRepository);
        return proofRepository.getAll(agentContext);
    }
    async findAllByQuery(agentContext, query) {
        const proofRepository = agentContext.dependencyManager.resolve(repository_1.ProofRepository);
        return proofRepository.findByQuery(agentContext, query);
    }
    /**
     * Find a proof record by id
     *
     * @param proofRecordId the proof record id
     * @returns The proof record or null if not found
     */
    findById(agentContext, proofRecordId) {
        const proofRepository = agentContext.dependencyManager.resolve(repository_1.ProofRepository);
        return proofRepository.findById(agentContext, proofRecordId);
    }
    async delete(agentContext, proofRecord, options) {
        var _a;
        const proofRepository = agentContext.dependencyManager.resolve(repository_1.ProofRepository);
        const didCommMessageRepository = agentContext.dependencyManager.resolve(didcomm_1.DidCommMessageRepository);
        await proofRepository.delete(agentContext, proofRecord);
        const deleteAssociatedDidCommMessages = (_a = options === null || options === void 0 ? void 0 : options.deleteAssociatedDidCommMessages) !== null && _a !== void 0 ? _a : true;
        if (deleteAssociatedDidCommMessages) {
            const didCommMessages = await didCommMessageRepository.findByQuery(agentContext, {
                associatedRecordId: proofRecord.id,
            });
            for (const didCommMessage of didCommMessages) {
                await didCommMessageRepository.delete(agentContext, didCommMessage);
            }
        }
    }
    /**
     * Retrieve a proof record by connection id and thread id
     *
     * @param properties Properties to query by
     *
     * @throws {RecordNotFoundError} If no record is found
     * @throws {RecordDuplicateError} If multiple records are found
     *
     * @returns The proof record
     */
    getByProperties(agentContext, properties) {
        const { threadId, connectionId, role } = properties;
        const proofRepository = agentContext.dependencyManager.resolve(repository_1.ProofRepository);
        return proofRepository.getSingleByQuery(agentContext, {
            connectionId,
            threadId,
            role,
        });
    }
    /**
     * Find a proof record by connection id and thread id, returns null if not found
     *
     * @param properties Properties to query by
     *
     * @returns The proof record
     */
    findByProperties(agentContext, properties) {
        const { role, connectionId, threadId } = properties;
        const proofRepository = agentContext.dependencyManager.resolve(repository_1.ProofRepository);
        return proofRepository.findSingleByQuery(agentContext, {
            connectionId,
            threadId,
            role,
        });
    }
    async update(agentContext, proofRecord) {
        const proofRepository = agentContext.dependencyManager.resolve(repository_1.ProofRepository);
        return await proofRepository.update(agentContext, proofRecord);
    }
}
exports.BaseProofProtocol = BaseProofProtocol;
