// @digitalcredentials/bitstring/dist/assertions.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/*!
 * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
 */
function isNumber(value, name) {
  if(typeof value !== 'number') {
    throw new TypeError(`"${name}" must be number.`);
  }
}

function isPositiveInteger(value, name) {
  if(!(Number.isInteger(value) && value > 0)) {
    throw new TypeError(`"${name}" must be a positive integer.`);
  }
}

function isString(value, name) {
  if(typeof value !== 'string') {
    throw new TypeError(`"${name}" must be a string.`);
  }
}

function isBoolean(value, name) {
  if(typeof value !== 'boolean') {
    throw new TypeError(`"${name}" must be a boolean.`);
  }
}

function isNonNegativeInteger(value, name) {
  if(!(Number.isInteger(value) && value >= 0)) {
    throw new TypeError(`"${name}" must be a non-negative integer.`);
  }
}

function isUint8Array(value, name) {
  if(!(value instanceof Uint8Array)) {
    throw new TypeError(`"${name}" must be a Uint8Array.`);
  }
}

exports.isBoolean = isBoolean;
exports.isNonNegativeInteger = isNonNegativeInteger;
exports.isNumber = isNumber;
exports.isPositiveInteger = isPositiveInteger;
exports.isString = isString;
exports.isUint8Array = isUint8Array;
